### DISCUSSION TITLE
Backwards Conversion from Spyder to Full Python File

### DISCUSSION THREAD

**Discussion by Charlie-Wong-Actuary** [fumitoh/modelx#80] [Q&A]
Hi all,

Sorry if I am posting a silly question. I have just been following the tutorial on Fibonacci Sequence, and have replicated the Python Code as following:

```python
# Section 0: Import Libraries
import modelx as mx
import pandas as pd

# Section 1: Creating Models & Space
SequenceModel = mx.new_model(name = "Sequence")
FiboSpace = SequenceModel.new_space(name = "Fibo")
param_df = pd.DataFrame({'n': [0]})
FiboSpace.param = param_df

# Section 2: Work with Fibo Space
@mx.defcells(FiboSpace)
def fibo(n):
    if n > 1:
        return fibo(n - 1) + fibo(n - 2)
    else:
        return n

@mx.defcells(FiboSpace)
def proj_len():
    return param['n'][0]

@mx.defcells(FiboSpace)
def result():
    return fibo(proj_len())

# Section n: Write Back the Model
mx.write_model(SequenceModel, r"SequenceModel")
```

# Question
By running the above coding, I am able to further develop in Spyder. In case, if I decided to continue the development in Python after changes in the Spyder Version, is there any method for me to convert the edited version in Spyder back to the Code in Python as above?

Thanks,
Charlie

**Comment by fumitoh** [comment]
If you want to keep updating the Python script as you develop the model in Spyder with the spyder-modelx plugin, here's how:

1. Open a new MxConsole
2. Create a new file (untitledN.py) in Editor by Ctrl+N
3. Paste the code above in untitledN.py, except for the last section (No need to write model here)
4. Run the file by F5, then the script runs in the MxConsole.

You can use the MxConsole and modelx widgets to explore the model.
Now, suppose you want to create another cells.
Append the definition of the cells to the file.

```python
@mx.defcells(FiboSpace)
def another():
    return 0
```
Then select the added lines, right-click and select 'Run selection or current line'.
Then the added cells is created.

Now, suppose you wan to change the definition of `result` as below:
```python
@mx.defcells(FiboSpace)
def result():
    return 2 * fibo(proj_len())
```
Change the definition in the editor, right-click and  'Run selection or current line'.
Now, you should find the new definition in the Properties tab in MxExplorer by double clicking on `result()` in the object tree.
This way you can keep updating your script while the model is updated and available in modelx widgets and the MxConsole.

There is another way. Spyder has a cool feature of separating a script into pieces ('cells') as Jupyter notebook does.
You can do so by writing comments starting with `# %%` [as explained in Spyder's manual](https://docs.spyder-ide.org/current/panes/editor.html#code-cells).
So, by changing the comment `# Section 1: Creating Models & Space` to `# %% Section 1: Creating Models & Space` and all other section comments, you can separate the entire script into cells and you can select one cell and run it by right-clicking and selecting 'Run cell'.
This is handy, for example,  when you make changes on multiple cells in Section 2 and want to update all of them at once.

**Comment by AADYABAJAJ** [comment]
Hi, a silly question - if I want to read the model as it is and just run it using IPython, what should I write?

Say, I want the function "result_pv" to be run from the model "BasicTerm_M" and results to be extracted out into a dataframe. What would be the code please?

**Reply by fumitoh** [reply]
See my answer to #86.
